File: 17-LetterCombinationsOfAPhoneNumber/17-LetterCombinationsOfAPhoneNumber.java
// Last updated: 8/16/2026, 9:53:33 PM
class Solution {
    private Map<Character, String> map = new HashMap<>();
    public List<String> letterCombinations(String digits) {
        // digits[i] will only be [2,9]
        List<String> res = new ArrayList<>();
        if(digits.length() == 0) return res;

        // get a map: digit - string
        map.put('2', "abc");
        map.put('3', "def");
        map.put('4', "ghi");
        map.put('5', "jkl");
        map.put('6', "mno");
        map.put('7', "pqrs");
        map.put('8', "tuv");
        map.put('9', "wxyz");

        // get all combinations -> DFS
        getCombinations(new StringBuilder(), digits, 0, res);
        
        return res;
    }

    // get all possible strings starting with 'curr', next available char are from digits[idx]
    private void getCombinations(StringBuilder curr, String digits, int idx, List<String> res) {

        if(idx == digits.length()) {
            res.add(curr.toString());
            return;
        }
        String chars = map.get(digits.charAt(idx));
        for(char c : chars.toCharArray()) {
            curr.append(c);
            getCombinations(curr, digits, idx + 1, res);
            curr.deleteCharAt(curr.length() - 1);
        }
    }

}

File: 17-LetterCombinationsofaPhoneNumber/17-LetterCombinationsofaPhoneNumber.py
# Last updated: 8/23/2026, 12:01:00 PM
1class Solution:
2    def letterCombinations(self, digits: str) -> List[str]:
3        # backtrack
4        if not digits:
5            return []
6        
7        phone = {
8            '2': 'abc', '3':'def', '4': 'ghi', '5': 'jkl', 
9            '6': 'mno', '7':'pqrs', '8': 'tuv', '9':'wxyz'
10        }
11        res = []
12        path = []   # curr path
13
14        # get legit path starting from digits[idx]
15        def backtrack(idx: int):
16            # exit: all digits used, then return
17            if len(path) == len(digits):
18                res.append(''.join(path))
19                return
20            
21            # latters mapping to the curr idx
22            letters = phone[digits[idx]]
23            for c in letters:
24                path.append(c)      # select letter from current digit
25                backtrack(idx + 1)  # process next digit
26                path.pop()          # cancel selection
27
28        backtrack(0)
29        return res